src/rust/perftools/profiler/mod.rs

```diff
@@ -85,121 +69,96 @@ pub fn set_callback(perf_callback: demi_callback_t) {
-
-                current.borrow_mut().add_succ(succ.clone());
-
-                succ
-            })
-        } else {
-            // We are currently not within any scope.
-            self.get_root_scope(name)
+    pub fn get_or_create_scope(&mut self, name: &'static str) -> Rc<RefCell<Scope>> {
+        match self.current_scope.as_ref() {
+            Some(current_scope) => {
+                let existing_scope = current_scope
+                    .borrow()
+                    .children_scopes
+                    .iter()
+                    .find(|child_scope| child_scope.borrow().name == name)
+                    .cloned();
+
+                existing_scope.unwrap_or_else(|| {
+                    let new_scope: Scope = Scope::new(name, Some(current_scope.clone()), self.perf_callback);
+                    let new_child_scope = Rc::new(RefCell::new(new_scope));
+
+                    current_scope.borrow_mut().add_child_scope(new_child_scope.clone());
+
+                    new_child_scope
+                })
+            },
+            None => self.get_or_create_root_scope(name),
         }
     }
 
-    /// Actually enter a scope.
     fn enter_scope(&mut self, scope: Rc<RefCell<Scope>>) -> Guard {
-        trace!("Entering scope: {}", scope.borrow().get_name());
-        let guard = scope.borrow_mut().enter();
-        self.current = Some(scope);
+        let guard = scope.borrow().enter();
+        self.current_scope = Some(scope);
 
         guard
     }
 
-    /// Completely reset profiling data.
     fn reset(&mut self) {
-        self.roots.clear();
+        self.root_scopes.clear();
 
         // Note that we could now still be anywhere in the previous profiling
         // tree, so we can not simply reset `self.current`. However, as the
         // frame comes to an end we will eventually leave a root node, at which
         // point `self.current` will be set to `None`.
     }
 
-    /// Leave the current scope.
     #[inline]
     fn leave_scope(&mut self, duration: u64) {
-        self.current = if let Some(current) = self.current.as_ref() {
-            cfg_if::cfg_if! {
-                if #[cfg(feature = "auto-calibrate")] {
-                    let d = duration.checked_sub(self.clock_drift);
-                    current.borrow_mut().leave(d.unwrap_or(duration));
-                } else {
-                    current.borrow_mut().leave(duration);
-                }
-            }
-
-            // Set current scope back to the parent node (if any).
-            current.borrow().get_pred().as_ref().cloned()
+        self.current_scope = if let Some(current_scope) = self.current_scope.as_ref() {
+            current_scope.borrow_mut().leave(duration);
+            current_scope.borrow().parent_scope.as_ref().cloned()
         } else {
             // This should not happen with proper usage.
             log::error!("Called perftools::profiler::leave() while not in any scope");
```

Comment:
Should this be a panic?

Reply:
I will send that as a separate change because if we change it to panic now, the tests will fali until we get all the timer/async_timer hierarchies right.